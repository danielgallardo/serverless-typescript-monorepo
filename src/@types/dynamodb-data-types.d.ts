/* tslint:disable */
declare module 'dynamodb-data-types' {
  export interface AttributeValue {
    B?: string;
    BS?: string[];
    BOOL?: boolean;
    L?: AttributeValue[];
    M?: {[id: string]: AttributeValue};
    N?: string;
    NS?: string[];
    NULL?: boolean;
    S?: string;
    SS?: string[];
  }

  export interface IAttributeValue {
    unwrap(obj: {[key: string]: AttributeValue}): {[key: string]: any};
    wrap(obj: {[key: string]: any}): {[key: string]: AttributeValue};
  }

  export const AttributeValue: IAttributeValue;
}
