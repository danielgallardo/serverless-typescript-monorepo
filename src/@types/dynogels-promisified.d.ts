declare module 'dynogels-promisified' {
  export import AWS = require('aws-sdk');
  import stream = require('stream');
  import {DocumentClient} from 'aws-sdk/lib/dynamodb/document_client';
  import * as joi from 'joi';
  import ConsumedCapacity = DocumentClient.ConsumedCapacity;

  // Dynogels Data Members
  export let log: Log;
  export let models: {[key: string]: Model};
  export let types: {
    stringSet(): joi.AnySchema;
    numberSet(): joi.AnySchema;
    binarySet(): joi.AnySchema;
    uuid(): joi.AnySchema;
    timeUUID(): joi.AnySchema;
  };

  export interface Log {
    info(...args: any[]): void;
    warn(...args: any[]): void;
  }

  // Dynogels global functions
  export function dynamoDriver(dynamoDB: AWS.DynamoDB): AWS.DynamoDB;

  export function reset(): void;

  export function define(modelName: string, config: ModelConfiguration): Model;

  export function createTables(callback: (err: string) => void): void;
  export function createTables(
    options: {[key: string]: CreateTablesOptions} | DynogelsGlobalOptions,
    callback: (err: string) => void
  ): void;

  export function Set(...args: any[]): any;

  export interface DynogelsGlobalOptions {
    $dynogels: {
      pollingInterval: number;
    };
  }

  export interface CreateTablesOptions {
    readCapacity?: number;
    writeCapacity?: number;
    streamSpecification?: {
      streamEnabled: boolean;
      streamViewType: string;
    };
  }

  export type LifeCycleAction = 'create' | 'update' | 'destroy';

  export type QueryResponse = {
    Items: Item[];
    LastEvaluatedKey?: object;
    ConsumedCapacity?: ConsumedCapacity;
    Count: number;
    ScannedCount: number;
  };

  // Dynogels Model
  export interface Model {
    new (attrs: {[key: string]: any}): Item;

    get(hashKey: any, rangeKey: any, options: GetItemOptions, callback: DynogelsItemCallback): void;
    get(haskKey: any, options: GetItemOptions, callback: DynogelsItemCallback): void;
    get(hashKey: any, callback: DynogelsItemCallback): void;
    get(hashKey: any, rangeKey: any, callback: DynogelsItemCallback): void;
    create(item: any, options: CreateItemOptions, callback: DynogelsItemCallback): void;
    create(item: any, callback: DynogelsItemCallback): void;
    update(item: any, options: UpdateItemOptions, callback: DynogelsItemCallback): void;
    update(item: any, callback: DynogelsItemCallback): void;
    destroy(
      hashKey: any,
      rangeKey: any,
      options: DestroyItemOptions,
      callback: DynogelsItemCallback
    ): void;
    destroy(haskKey: any, options: DestroyItemOptions, callback: DynogelsItemCallback): void;
    destroy(hashKey: any, callback: DynogelsItemCallback): void;
    destroy(hashKey: any, rangeKey: any, callback: DynogelsItemCallback): void;
    destroy(item: any, options: DestroyItemOptions, callback: DynogelsItemCallback): void;
    destroy(item: any, callback: DynogelsItemCallback): void;
    query(hashKey: any): Query;
    scan(): Scan;
    parallelScan(totalSegments: number): Scan;
    getItems(
      items: string[] | Array<{[key: string]: string}>,
      callback: (err: Error, items: any[]) => void
    ): void;
    getItems(
      items: string[] | Array<{[key: string]: string}>,
      options: GetItemOptions,
      callback: (err: Error, items: any[]) => void
    ): void;
    batchGetItems(
      items: string[] | Array<{[key: string]: string}>,
      callback: (err: Error, items: any[]) => void
    ): void;
    batchGetItems(
      items: string[] | Array<{[key: string]: string}>,
      options: GetItemOptions,
      callback: (err: Error, items: any[]) => void
    ): void;
    createTable(
      options: CreateTablesOptions,
      callback: (err: Error, data: AWS.DynamoDB.CreateTableOutput) => void
    ): void;
    createTable(callback: (err: Error, data: AWS.DynamoDB.CreateTableOutput) => void): void;
    updateTable(
      throughput: Throughput,
      callback: (err: Error, data: AWS.DynamoDB.UpdateTableOutput) => void
    ): void;
    updateTable(callback: (err: Error, data: AWS.DynamoDB.UpdateTableOutput) => void): void;
    describeTable(callback: (err: Error, data: AWS.DynamoDB.DescribeTableOutput) => void): void;
    deleteTable(callback: (err: Error) => void): void;
    tableName(): string;

    after(action: LifeCycleAction, listner: (item: Item) => void): void;
    before(
      action: LifeCycleAction,
      listner: (data: any, next: (err: Error | null, data: any) => void) => void
    ): void;
    config(config: ModelConfig): {name: string};

    getAsync(hashKey: any, rangeKey: any, options: GetItemOptions): Promise<Item>;
    getAsync(haskKey: any, options: GetItemOptions): Promise<Item>;
    getAsync(hashKey: any): Promise<Item>;
    getAsync(hashKey: any, rangeKey: any): Promise<Item>;
    createAsync(item: any, options: CreateItemOptions): Promise<Item>;
    createAsync(item: any): Promise<Item>;
    updateAsync(item: any, options: UpdateItemOptions): Promise<Item>;
    updateAsync(item: any): Promise<Item>;
    destroyAsync(hashKey: any, rangeKey: any, options: DestroyItemOptions): Promise<Item>;
    destroyAsync(haskKey: any, options: DestroyItemOptions): Promise<Item>;
    destroyAsync(hashKey: any): Promise<Item>;
    destroyAsync(hashKey: any, rangeKey: any): Promise<Item>;
    destroyAsync(item: any, options: DestroyItemOptions): Promise<Item>;
    destroyAsync(item: any): Promise<Item>;
    parallelScanAsync(totalSegments: number): Scan;
    getItemsAsync(items: string[] | Array<{[key: string]: string}>): Promise<any[]>;
    getItemsAsync(
      items: string[] | Array<{[key: string]: string}>,
      options: GetItemOptions
    ): Promise<any[]>;
    batchGetItemsAsync(items: string[] | Array<{[key: string]: string}>): Promise<any[]>;
    batchGetItemsAsync(
      items: string[] | Array<{[key: string]: string}>,
      options: GetItemOptions
    ): Promise<any[]>;
    createTableAsync(options: CreateTablesOptions): Promise<AWS.DynamoDB.CreateTableOutput>;
    createTableAsync(): Promise<AWS.DynamoDB.CreateTableOutput>;
    updateTableAsync(throughput: Throughput): Promise<AWS.DynamoDB.UpdateTableOutput>;
    updateTableAsync(): Promise<AWS.DynamoDB.UpdateTableOutput>;
    describeTableAsync(): Promise<AWS.DynamoDB.DescribeTableOutput>;
    deleteTableAsync(): Promise<any>;
  }

  export type DynogelsItemCallback = (err: Error, data: Item) => void;

  export interface Throughput {
    readCapacity: number;
    writeCapacity: number;
  }

  export interface CreateItemOptions {
    expected?: {[key: string]: any};
    overwrite?: boolean;

    Expected?: AWS.DynamoDB.ExpectedAttributeMap;
    ReturnValues?: AWS.DynamoDB.ReturnValue;
    ReturnConsumedCapacity?: AWS.DynamoDB.ReturnConsumedCapacity;
    ReturnItemCollectionMetrics?: AWS.DynamoDB.ReturnItemCollectionMetrics;
    ConditionalOperator?: AWS.DynamoDB.ConditionalOperator;
    ConditionExpression?: AWS.DynamoDB.ConditionExpression;
    ExpressionAttributeNames?: AWS.DynamoDB.ExpressionAttributeNameMap;
    ExpressionAttributeValues?: {[key: string]: any};
  }

  export interface UpdateItemOptions {
    expected?: {[key: string]: any};

    AttributeUpdates?: AWS.DynamoDB.AttributeUpdates;
    Expected?: AWS.DynamoDB.ExpectedAttributeMap;
    ConditionalOperator?: AWS.DynamoDB.ConditionalOperator;
    ReturnValues?: AWS.DynamoDB.ReturnValue;
    ReturnConsumedCapacity?: AWS.DynamoDB.ReturnConsumedCapacity;
    ReturnItemCollectionMetrics?: AWS.DynamoDB.ReturnItemCollectionMetrics;
    UpdateExpression?: AWS.DynamoDB.UpdateExpression;
    ConditionExpression?: AWS.DynamoDB.ConditionExpression;
    ExpressionAttributeNames?: AWS.DynamoDB.ExpressionAttributeNameMap;
    ExpressionAttributeValues?: {[key: string]: any};
  }

  export interface DestroyItemOptions {
    Expected?: AWS.DynamoDB.ExpectedAttributeMap;
    ConditionalOperator?: AWS.DynamoDB.ConditionalOperator;
    ReturnValues?: AWS.DynamoDB.ReturnValue;
    ReturnConsumedCapacity?: AWS.DynamoDB.ReturnConsumedCapacity;
    ReturnItemCollectionMetrics?: AWS.DynamoDB.ReturnItemCollectionMetrics;
    ConditionExpression?: AWS.DynamoDB.ConditionExpression;
    ExpressionAttributeNames?: AWS.DynamoDB.ExpressionAttributeNameMap;
    ExpressionAttributeValues?: {[key: string]: any};
  }

  export interface GetItemOptions {
    AttributesToGet?: AWS.DynamoDB.AttributeNameList;
    ConsistentRead?: AWS.DynamoDB.ConsistentRead;
    ReturnConsumedCapacity?: AWS.DynamoDB.ReturnConsumedCapacity;
    ProjectionExpression?: AWS.DynamoDB.ProjectionExpression;
    ExpressionAttributeNames?: AWS.DynamoDB.ExpressionAttributeNameMap;
  }

  export interface ModelConfig {
    tableName?: string;
    docClient?: any;
    dynamodb?: AWS.DynamoDB;
  }

  // Dynogels Item
  export interface Item {
    get(): {[key: string]: any};
    get(key: string): any;
    set(params: {}): Item;
    save(callback?: DynogelsItemCallback): void;
    update(options: UpdateItemOptions, callback?: DynogelsItemCallback): void;
    update(callback?: DynogelsItemCallback): void;
    destroy(options: DestroyItemOptions, callback?: DynogelsItemCallback): void;
    destroy(callback?: DynogelsItemCallback): void;
    toJSON(): any;
    toPlainObject(): any;
  }

  export interface BaseChain<T> {
    equals(value: any): T;
    eq(value: any): T;
    lte(value: any): T;
    lt(value: any): T;
    gte(value: any): T;
    gt(value: any): T;
    null(): T;
    exists(): T;
    beginsWith(value: any): T;
    between(value1: any, value2: any): T;
  }

  export interface ExtendedChain<T> extends BaseChain<T> {
    contains(value: any): T;
    notContains(value: any): T;
    in(values: any[]): T;
    ne(value: any): T;
  }

  export type QueryWhereChain = BaseChain<Query>;
  export type QueryFilterChain = ExtendedChain<Query>;

  // Dynogels Query
  export interface Query {
    limit(number: number): Query;
    filterExpression(expression: any): Query;
    expressionAttributeNames(data: any): Query;
    expressionAttributeValues(data: any): Query;
    projectionExpression(data: any): Query;
    usingIndex(name: string): Query;
    consistentRead(read: boolean): Query;
    addKeyCondition(condition: any): Query;
    addFilterCondition(condition: any): Query;
    startKey(hashKey: any, rangeKey?: any): Query;
    attributes(attrs: any): Query;
    ascending(): Query;
    descending(): Query;
    select(value: any): Query;
    returnConsumedCapacity(value: any): Query;
    loadAll(): Query;
    where(keyName: string): QueryWhereChain;
    filter(keyName: string): QueryFilterChain;
    exec(): stream.Readable;
    exec(callback: (err: Error, data: QueryResponse) => void): void;
    execAsync(): Promise<QueryResponse>;
  }

  export interface ScanWhereChain extends ExtendedChain<Scan> {
    notNull(): Scan;
  }

  // Dynogels Scan
  export interface Scan {
    limit(number: number): Scan;
    addFilterCondition(condition: any): Scan;
    startKey(hashKey: any, rangeKey?: any): Scan;
    attributes(attrs: any): Scan;
    select(value: any): Scan;
    returnConsumedCapacity(): Scan;
    segments(segment: any, totalSegments: number): Scan;
    where(keyName: string): ScanWhereChain;
    filterExpression(expression: any): Scan;
    expressionAttributeNames(data: any): Scan;
    expressionAttributeValues(data: any): Scan;
    projectionExpression(data: any): Scan;
    exec(): stream.Readable;
    exec(callback: (err: Error, data: QueryResponse) => void): void;
    loadAll(): Scan;
    execAsync(): Promise<QueryResponse>;
  }

  export type tableResolve = () => string;

  export interface SchemaType {
    [key: string]: joi.AnySchema | SchemaType;
  }

  export interface ModelConfiguration {
    hashKey: string;
    rangeKey?: string;
    timestamps?: boolean;
    createdAt?: boolean;
    updatedAt?: string;
    schema?: SchemaType;
    validation?: joi.ValidationOptions;
    tableName?: string | tableResolve;
    indexes?: any[];
    log?: Log;
  }

  export interface Document {
    [key: string]: any;
  }

  export interface DocumentCollection {
    Items: Document[];
    Count: number;
    ScannedCount: number;
  }
}
