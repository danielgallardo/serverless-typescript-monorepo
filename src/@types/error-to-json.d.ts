declare module 'error-to-json' {
  export type JSONError = {
    name: string;
    message: string;
    stack?: string[];
    [key: string]: any;
  };

  function errorToJSON(error: Error): JSONError;

  export default errorToJSON;
}
