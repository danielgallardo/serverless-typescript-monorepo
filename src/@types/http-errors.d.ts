declare module 'http-errors' {
  interface IHttpError extends Error {
    status: number;
    statusCode: number;
    expose: boolean;
    headers?: {
      [key: string]: string;
    };
    [key: string]: any;
  }

  type HttpErrorConstructor = new (msg?: string) => IHttpError;

  type CreateHttpError = (
    ...args: Array<Error | string | number | {[key: string]: any}>
  ) => IHttpError;

  type NamedConstructors = {
    [code: string]: HttpErrorConstructor;
    HttpError: HttpErrorConstructor;
    BadRequest: HttpErrorConstructor;
    Unauthorized: HttpErrorConstructor;
    PaymentRequired: HttpErrorConstructor;
    Forbidden: HttpErrorConstructor;
    NotFound: HttpErrorConstructor;
    MethodNotAllowed: HttpErrorConstructor;
    NotAcceptable: HttpErrorConstructor;
    ProxyAuthenticationRequired: HttpErrorConstructor;
    RequestTimeout: HttpErrorConstructor;
    Conflict: HttpErrorConstructor;
    Gone: HttpErrorConstructor;
    LengthRequired: HttpErrorConstructor;
    PreconditionFailed: HttpErrorConstructor;
    PayloadTooLarge: HttpErrorConstructor;
    URITooLong: HttpErrorConstructor;
    UnsupportedMediaType: HttpErrorConstructor;
    RangeNotSatisfiable: HttpErrorConstructor;
    ExpectationFailed: HttpErrorConstructor;
    ImATeapot: HttpErrorConstructor;
    MisdirectedRequest: HttpErrorConstructor;
    UnprocessableEntity: HttpErrorConstructor;
    Locked: HttpErrorConstructor;
    FailedDependency: HttpErrorConstructor;
    UnorderedCollection: HttpErrorConstructor;
    UpgradeRequired: HttpErrorConstructor;
    PreconditionRequired: HttpErrorConstructor;
    TooManyRequests: HttpErrorConstructor;
    RequestHeaderFieldsTooLarge: HttpErrorConstructor;
    UnavailableForLegalReasons: HttpErrorConstructor;
    InternalServerError: HttpErrorConstructor;
    NotImplemented: HttpErrorConstructor;
    BadGateway: HttpErrorConstructor;
    ServiceUnavailable: HttpErrorConstructor;
    GatewayTimeout: HttpErrorConstructor;
    HTTPVersionNotSupported: HttpErrorConstructor;
    VariantAlsoNegotiates: HttpErrorConstructor;
    InsufficientStorage: HttpErrorConstructor;
    LoopDetected: HttpErrorConstructor;
    BandwidthLimitExceeded: HttpErrorConstructor;
    NotExtended: HttpErrorConstructor;
    NetworkAuthenticationRequire: HttpErrorConstructor;
  };

  const createHttpError: CreateHttpError & NamedConstructors;
  export = createHttpError;
}
