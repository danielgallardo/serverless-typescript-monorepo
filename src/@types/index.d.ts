import {APIGatewayEventRequestContext} from 'aws-lambda';

export interface INormalizedEvent {
  headers: {
    [name: string]: string;
  };
  httpMethod: string;
  isBase64Encoded: boolean;
  path: string;
  pathParameters: {[name: string]: string} | {};
  queryStringParameters: {[name: string]: string} | {};
  stageVariables: {[name: string]: string} | null;
  requestContext: APIGatewayEventRequestContext;
  resource: string;
  body: any;
}
