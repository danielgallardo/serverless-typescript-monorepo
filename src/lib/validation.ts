import env from 'env-var';
import _Joi, {Root, SchemaLike, StringSchema, ValidationOptions} from 'joi';
import {JoiError} from './errors';
import {errorToJSON, logDebugInfo, logError, logInfo} from './logger';

export interface IJoi extends Root {
  /**
   * Use this validation for queryParameters and pathParameters
   * to avoid undefined and null strings passed from front-end
   */
  param(): this;
  /**
   * Use this validation for unix timestamps in milliseconds
   * it will not convert value to date
   * use Joi.date().timestamp() to automatically convert value to date
   */
  timestamp(): this;
}

export interface ValidationConfig extends ValidationOptions {
  internal?: boolean;
}

export const Joi = _Joi.extend([
  {
    name: 'param',
    base: _Joi
      .string()
      .not('undefined', 'null')
      .required()
  },
  {
    name: 'timestamp',
    base: _Joi
      .date()
      .timestamp()
      .raw()
  }
]) as IJoi;

/**
 * Attempts to return a valid value, if value is invalid throws a JoiError
 * @param value - a value to validate
 * @param schema - Joi schema
 * @param errorPrefix - a sting that is prepended to error message
 * @param options - Joi validation options
 */
export function attempt<T>(
  value: T,
  schema: SchemaLike,
  errorPrefix?: string,
  options: ValidationConfig = {}
): T {
  const {internal, ...joiOptions} = options;
  const result = Joi.validate<T>(value, schema, joiOptions);
  if (result.error) {
    throw new JoiError(result.error, {errorPrefix, internal});
  }
  return result.value;
}

/**
 * Asserts that value is valid, if value is invalid throws a JoiError
 * @param value - a value to validate
 * @param schema - Joi schema
 * @param errorPrefix - a sting that is prepended to error message
 * @param options - Joi validation options
 */
export function assert<T>(
  value: T,
  schema: SchemaLike,
  errorPrefix?: string,
  options: ValidationConfig = {}
): void {
  const {internal, ...joiOptions} = options;
  const result = Joi.validate<T>(value, schema, joiOptions);
  if (result.error) {
    throw new JoiError(result.error, {errorPrefix, internal});
  }
}

/**
 * Gets and validates process.env variable
 * @param name
 * @param defaultValue
 */
export function getEnv(name: string, defaultValue?: string): string {
  try {
    if (defaultValue) return env.get(name, defaultValue).asString();
    return env
      .get(name)
      .required()
      .asString();
  } catch (error) {
    // we need to log an error right away in case getEnv is used outside lambda handler
    // otherwise we'll get only "module initialization error: EnvVarError"
    console.log(`== MISSING ENV VARIABLE: "${name}" ==`, JSON.stringify(errorToJSON(error)));
    throw error;
  }
}
