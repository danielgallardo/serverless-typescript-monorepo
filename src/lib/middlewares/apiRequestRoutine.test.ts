import {APIGatewayProxyEvent} from 'aws-lambda';
import clone from 'clone';
import {NotFound, UnprocessableEntity} from 'http-errors';
import middy from 'middy';
import {INormalizedEvent} from '../../@types';
import {event} from '../fixtures/events';
import {Joi} from '../validation';
import {
  apiRequestRoutine,
  buildErrorResponse,
  buildJSONResponse,
  parseJSONBody,
  setDefaultParams
} from './apiRequestRoutine';

describe('apiRequestRoutine', () => {
  let demoEvent: APIGatewayProxyEvent;
  beforeEach(() => {
    demoEvent = clone(event);
  });
  it('parseJSONBody should return convert event.body to json', () => {
    parseJSONBody(demoEvent);
    expect(demoEvent.body).toEqual({foo: 'bar'});
  });

  it('parseJSONBody should throw error if event.body is not a json', () => {
    demoEvent.body = 'not a json';
    expect(parseJSONBody.bind(null, demoEvent)).toThrowError(UnprocessableEntity);
  });

  it('parseJSONBody should do nothing if there is no body', () => {
    demoEvent.body = null;
    parseJSONBody(demoEvent);
    expect(demoEvent.body).toBeNull();
  });

  it('setDefaultParams should set default query and path parameters', () => {
    setDefaultParams(demoEvent);
    expect(demoEvent.queryStringParameters).toEqual({});
    expect(demoEvent.pathParameters).toEqual({});
  });

  it('buildJSONResponse should build a correct response object', () => {
    expect(buildJSONResponse({foo: 'bar'})).toMatchSnapshot();
  });

  it('buildJSONResponse should build redirect response if redirectUrl is present', () => {
    expect(buildJSONResponse({redirectUrl: 'https://example.com'})).toMatchSnapshot();
  });

  it('buildJSONResponse should build custom response if statusCode is present', () => {
    expect(buildJSONResponse({statusCode: 200, body: 'custom body'})).toMatchSnapshot();
  });

  it('buildJSONResponse should build no content if response is empty', () => {
    expect(buildJSONResponse(undefined)).toMatchSnapshot();
  });

  it('buildErrorResponse should build a correct response object in case of the internal error', () => {
    expect(buildErrorResponse(new Error('Some internal error'))).toMatchSnapshot();
  });

  it('buildErrorResponse should build a correct response object in case of the validation error', () => {
    const schema = Joi.object().keys({foo: Joi.string().required()});
    const {error} = Joi.validate({}, schema);
    expect(buildErrorResponse(error)).toMatchSnapshot();
  });

  it('buildErrorResponse should build a correct response object in case of the http error', () => {
    expect(buildErrorResponse(new NotFound('Recourse not found'))).toMatchSnapshot();
  });

  it('should normalize event', () => {
    const handler = middy(async (event: INormalizedEvent) => {
      expect(event).toMatchSnapshot();
    });
    handler.use(apiRequestRoutine());
    // @ts-ignore
    handler(demoEvent, {}, () => {});
  });

  it('should return success response', () => {
    const handler = middy(async () => {
      return {foo: 'bar'};
    });
    handler.use(apiRequestRoutine());

    // @ts-ignore
    handler(demoEvent, {}, (error, response) => {
      expect(error).toBeNull();
      expect(response).toMatchSnapshot();
    });
  });

  it('should return error response if http error is thrown in handler', () => {
    const handler = middy(async () => {
      throw new NotFound('Recourse not found');
    });
    handler.use(apiRequestRoutine());

    // @ts-ignore
    handler(demoEvent, {}, (error, response) => {
      expect(error).toBeNull();
      expect(response).toMatchSnapshot();
    });
  });

  it('should return error response if internal error is thrown in handler', () => {
    const handler = middy(async () => {
      throw new Error('Some internal error');
    });
    handler.use(apiRequestRoutine());

    // @ts-ignore
    handler(demoEvent, {}, (error, response) => {
      expect(error).toBeNull();
      expect(response).toMatchSnapshot();
    });
  });
});
