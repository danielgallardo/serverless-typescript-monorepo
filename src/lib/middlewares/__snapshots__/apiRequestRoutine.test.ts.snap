// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apiRequestRoutine buildErrorResponse should build a correct response object in case of the http error 1`] = `
Object {
  "body": "{\\"status\\":\\"NotFoundError\\",\\"statusCode\\":404,\\"message\\":\\"Recourse not found\\"}",
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
  },
  "statusCode": 404,
}
`;

exports[`apiRequestRoutine buildErrorResponse should build a correct response object in case of the internal error 1`] = `
Object {
  "body": "{\\"status\\":\\"ServiceUnavailableError\\",\\"statusCode\\":503,\\"message\\":\\"Service Unavailable\\"}",
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
  },
  "statusCode": 503,
}
`;

exports[`apiRequestRoutine buildErrorResponse should build a correct response object in case of the validation error 1`] = `
Object {
  "body": "{\\"status\\":\\"UnprocessableEntityError\\",\\"statusCode\\":422,\\"message\\":\\"\\\\\\"foo\\\\\\" is required\\",\\"details\\":[{\\"message\\":\\"\\\\\\"foo\\\\\\" is required\\",\\"path\\":[\\"foo\\"],\\"type\\":\\"any.required\\",\\"context\\":{\\"key\\":\\"foo\\",\\"label\\":\\"foo\\"}}]}",
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
  },
  "statusCode": 422,
}
`;

exports[`apiRequestRoutine buildJSONResponse should build a correct response object 1`] = `
Object {
  "body": "{\\"foo\\":\\"bar\\"}",
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
  },
  "statusCode": 200,
}
`;

exports[`apiRequestRoutine buildJSONResponse should build custom response if statusCode is present 1`] = `
Object {
  "body": "custom body",
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
  },
  "statusCode": 200,
}
`;

exports[`apiRequestRoutine buildJSONResponse should build no content if response is empty 1`] = `
Object {
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
  },
  "statusCode": 204,
}
`;

exports[`apiRequestRoutine buildJSONResponse should build redirect response if redirectUrl is present 1`] = `
Object {
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
    "Location": "https://example.com",
  },
  "statusCode": 302,
}
`;

exports[`apiRequestRoutine should normalize event 1`] = `
Object {
  "body": Object {
    "foo": "bar",
  },
  "headers": Object {
    "Accept": "application/json",
    "Content-Type": "application/json",
  },
  "httpMethod": "POST",
  "isBase64Encoded": false,
  "multiValueHeaders": Object {
    "foo": Array [
      "bar",
    ],
  },
  "multiValueQueryStringParameters": Object {
    "foo": Array [
      "bar",
    ],
  },
  "path": "/test",
  "pathParameters": Object {},
  "queryStringParameters": Object {},
  "requestContext": Object {
    "accountId": "862946620411",
    "apiId": "q27n90kff1",
    "httpMethod": "POST",
    "identity": Object {
      "accessKey": null,
      "accountId": null,
      "apiKey": null,
      "apiKeyId": null,
      "caller": null,
      "cognitoAuthenticationProvider": null,
      "cognitoAuthenticationType": null,
      "cognitoIdentityId": null,
      "cognitoIdentityPoolId": null,
      "sourceIp": "35.123.74.111",
      "user": null,
      "userAgent": "axios/0.15.2",
      "userArn": null,
    },
    "path": "/dev/test",
    "requestId": "c901f211-cc5e-11e8-b7a5-4bcd36c6e6f3",
    "requestTimeEpoch": 1539156810614,
    "resourceId": "una98p",
    "resourcePath": "/test",
    "stage": "dev",
  },
  "resource": "/test",
  "stageVariables": null,
}
`;

exports[`apiRequestRoutine should return error response if http error is thrown in handler 1`] = `
Object {
  "body": "{\\"status\\":\\"NotFoundError\\",\\"statusCode\\":404,\\"message\\":\\"Recourse not found\\"}",
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
  },
  "statusCode": 404,
}
`;

exports[`apiRequestRoutine should return error response if internal error is thrown in handler 1`] = `
Object {
  "body": "{\\"status\\":\\"ServiceUnavailableError\\",\\"statusCode\\":503,\\"message\\":\\"Service Unavailable\\"}",
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
  },
  "statusCode": 503,
}
`;

exports[`apiRequestRoutine should return success response 1`] = `
Object {
  "body": "{\\"foo\\":\\"bar\\"}",
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
  },
  "statusCode": 200,
}
`;
